pkg/csi_driver/controller.go

```diff
@@ -101,15 +106,61 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
 	} else {
+
+		// Get the reserved-IPV4-cidr from the request. It was not added in the service instance as it is only consumed by the controller
+		reservedIPV4CIDR, reservedIPV4CIDRPresent := req.GetParameters()[paramReservedIPV4CIDR]
+		if !reservedIPV4CIDRPresent {
+			return nil, fmt.Errorf("The required parameter reserved-ipv4-cidr was not provided")
+		}
+
+		// Prevents lock acquisition in case the CIDR is invalid
+		err = s.config.ipAllocator.ValidateCIDR(reservedIPV4CIDR)
+		if err != nil {
+			return nil, err
+		}
+
+		reservedIPRange, err := s.getUnreservedIPBlock(ctx, newFiler, reservedIPV4CIDR)
+		if err != nil {
+			return nil, err
+		}
+
+		// Adding the reserved IP range to the instance object. It is freed up when the instance is created as we would then be getting this value in the list instances response for subsequent requests
+		newFiler.Network.ReservedIpRange = reservedIPRange
+
 		// Create the instance
 		filer, err = s.config.fileService.CreateInstance(ctx, newFiler)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 	}
+
+	//Free up the reserved IPV4 Range from the ipAllocator as we would be recieving this value in the listInstances call as the createInstance operation was completed
+	s.config.ipAllocator.Unreserve(filer.Network.ReservedIpRange)
```

Comment:
Are we reserving the IP somewhere? Why reserve and unreserve in the same call?

Reply:
We are reserving in the getUnreservedIPBlock() function which is the inside a lock. It is unreserved when your CreateInstance call resulted in a success.

Reply:
It's a little unintuitive why you would unreserve an IP once creation is successful.

Maybe better names could be Hold/Unhold.